File: 3534-count-almost-equal-pairs-i/3534-count-almost-equal-pairs-i.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public int countPairs(int[] nums) {
        int n = nums.length;
        int ans = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isAlmostEqual(nums[i], nums[j])) {
                    ans++;
                }
            }
        }
        
        return ans;
    }

    private boolean isAlmostEqual(int x, int y) {
        String s1 = Integer.toString(x);
        String s2 = Integer.toString(y);
        
        // Adjust lengths by padding with leading zeros
        while (s1.length() < s2.length()) {
            s1 = "0" + s1;
        }
        while (s2.length() < s1.length()) {
            s2 = "0" + s2;
        }

        int diffCount = 0;
        List<Integer> diffIndex = new ArrayList<>();
        
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                diffCount++;
                diffIndex.add(i);
                if (diffCount > 2) {
                    return false;
                }
            }
        }
        
        if (diffCount == 0) {
            return true;
        }
        if (diffCount == 2) {
            // Ensure index access is within bounds
            char[] s1Array = s1.toCharArray();
            int idx1 = diffIndex.get(0);
            int idx2 = diffIndex.get(1);

            // Swap characters
            char temp = s1Array[idx1];
            s1Array[idx1] = s1Array[idx2];
            s1Array[idx2] = temp;

            return String.valueOf(s1Array).equals(s2);
        }
        
        return false;
    }
} 

File: 3534-count-almost-equal-pairs-i/3534-count-almost-equal-pairs-i.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countPairs = function(nums) {
    let n = nums.length;
    let ans = 0;
    
    // Loop through each pair (i, j) where i < j
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            if (isAlmostEqual(nums[i], nums[j])) {
                ans++;
            }
        }
    }
    
    return ans;
};

/**
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
function isAlmostEqual(x, y) {
    let s1 = x.toString();
    let s2 = y.toString();
    
    // Adjust lengths by padding with leading zeros
    while (s1.length < s2.length) {
        s1 = "0" + s1;
    }
    while (s2.length < s1.length) {
        s2 = "0" + s2;
    }

    let diffCount = 0;
    let diffIndex = [];
    
    // Compare characters of both numbers
    for (let i = 0; i < s1.length; i++) {
        if (s1[i] !== s2[i]) {
            diffCount++;
            diffIndex.push(i);
            if (diffCount > 2) {
                return false;  // More than 2 differences, not almost equal
            }
        }
    }
    
    // If there are no differences, they are equal
    if (diffCount === 0) {
        return true;
    }
    
    // If there are exactly 2 differences, check if swapping makes them equal
    if (diffCount === 2) {
        let s1Array = s1.split('');
        let idx1 = diffIndex[0];
        let idx2 = diffIndex[1];

        // Swap characters
        let temp = s1Array[idx1];
        s1Array[idx1] = s1Array[idx2];
        s1Array[idx2] = temp;

        return s1Array.join('') === s2;
    }
    
    return false;
}
